apps/concierge_site/lib/controllers/rejected_email_controller.ex

```diff
@@ -4,35 +4,84 @@ defmodule ConciergeSite.RejectedEmailController do
   alias AlertProcessor.Model.User
   require Logger
 
-  def handle_rejected_email(conn, params) do
-    do_handle_rejected_email(conn, params)
+  @ex_aws Application.fetch_env!(:alert_processor, :ex_aws)
+
+  def handle_message(conn, _params) do
+    with {:ok, raw_body, conn} <- read_body(conn),
+         {:ok, params} <- Poison.decode(raw_body),
+         :ok <- @ex_aws.SNS.verify_message(params) do
+      do_handle_message(params)
+      send_resp(conn, :no_content, "")
+    else
+      {:more, partial_body, conn} ->
+        log_error("event=handle_error reason=too_large #{partial_body}")
+        send_resp(conn, :request_entity_too_large, "")
+
+      {:error, {:invalid, _}} ->
+        log_error("event=handle_error reason=invalid_json")
+        send_resp(conn, :bad_request, "")
+
+      {:error, error} ->
+        log_error("event=handle_error reason=verify #{error}")
+        send_resp(conn, :unauthorized, "")
+    end
+  end
+
+  defp do_handle_message(%{"Type" => "SubscriptionConfirmation", "SubscribeURL" => url}) do
+    {:ok, %{status_code: 200}} = HTTPoison.get(url)
+    log_info("event=subscription_confirmation")
+  end
+
+  # No action required for these
+  defp do_handle_message(%{"Type" => "UnsubscribeConfirmation"}),
+    do: log_info("event=unsubscribe_confirmation")
+
+  defp do_handle_message(%{"Type" => "Notification", "Message" => message_json}) do
+    message = Poison.decode!(message_json)
+    log_info("event=notification #{inspect(message)}")
+    handle_notification(message)
   end
 
-  defp do_handle_rejected_email(conn, %{"notificationType" => "Bounce", "bounce" => params}) do
-    log_rejected_email(params["bouncedRecipients"])
+  defp handle_notification(%{"notificationType" => "Bounce", "bounce" => bounce}),
+    do: handle_bounce(bounce)
+
+  defp handle_notification(%{"notificationType" => "Complaint", "complaint" => complaint}),
+    do: handle_complaint(complaint)
 
-    conn
-    |> put_status(:ok)
-    |> json(%{})
+  # Since we only send emails to single recipients, assert that reports are only for a single
+  # recipient. This is important for complaints, since if the list contains multiple recipients,
+  # it's not guaranteed that complaints were actually received from all of them. See:
+  # https://docs.aws.amazon.com/ses/latest/DeveloperGuide/notification-contents.html#complained-recipients
+
+  defp handle_bounce(%{
+         "bounceType" => "Permanent",
+         "bouncedRecipients" => [%{"emailAddress" => email}]
+       }) do
+    get_and_update_user(email, &User.set_email_rejection(&1, "bounce"))
   end
 
-  defp do_handle_rejected_email(conn, %{"notificationType" => "Complaint", "complaint" => params}) do
-    log_rejected_email(params["complainedRecipients"])
+  # Take no action on other bounce types, but assert the type is one of the expected ones
+  defp handle_bounce(%{"bounceType" => type}) when type in ~w(Transient Undetermined), do: nil
```

Comment:
Do we want to consider logging these just for awareness or the ability to look them up later?

Reply:
We are logging all notifications on L41, so that will catch these as well.

Reply:
Oops, yep missed that 👍